app/fastcgi_cert_manager.c

```diff
@@ -75,120 +74,139 @@ get_value_from_query_list(const char *search_key, UriQueryListA *queryList)
 }
 
 gchar *
-write_file_from_stream(int contentLength, FCGX_Request request)
+write_file_from_stream(const char *cert_file_name, int contentLength, FCGX_Request request)
 {
+  bool success = false;
+  const char *MULTIPART_FORM_DATA = "multipart/form-data";
   gchar *file_path = NULL;
+  FCGI_FILE *fileOut = NULL;
   const char *datastart = "\r\n\r\n";
   char *contentType = FCGX_GetParam("CONTENT_TYPE", request.envp);
-  syslog(LOG_INFO, "ContentType %s", contentType);
-  char *boundarytext =
-      strstr(contentType, "boundary="); // Use to find the endof input
-  boundarytext += strlen("boundary=");
-  syslog(LOG_INFO, "Boundary text %s%s%s", "SB>", boundarytext, "<EB");
-
-  // create at tmpdir for the file
-  if (tmp_file_dir == NULL) {
-    tmp_file_dir = g_strdup_printf("%s%s", app_localdata_path, "tmp_upload");
-
-    if (g_mkdir_with_parents(tmp_file_dir, 740) != 0) { // TODO Check permission
-      syslog(LOG_ERR,
-             "Failed to create temporary directory:%s, error: %s",
-             tmp_file_dir,
-             strerror(errno));
-      tmp_file_dir = NULL;
+
+  syslog(LOG_INFO, "Content-Type: %s", contentType);
+  if (!strncmp(contentType, MULTIPART_FORM_DATA, sizeof(MULTIPART_FORM_DATA)-1)) {
+    char *boundarytext =
+        strstr(contentType, "boundary="); /* Use to find the endof input */
+    boundarytext += strlen("boundary=");
+    syslog(LOG_INFO, "Boundary text %s%s%s", "SB>", boundarytext, "<EB");
+
+    /* TODO: Check for FILE_EXISTS and retry */
+    file_path = g_strdup_printf("%s/%s.%06d.tmp", app_localdata_path, cert_file_name, g_random_int_range(0, 999999));
```

Comment:
check mkdtemp() and related functions to not have to create the name yourself

Reply:
To use mkstemp() would mean moving abaonding the FCGI file fundling functionality through out I think. Is that what you would prefer? I'm not sure that it would offer any real advantage over the current strategy..